server/src/main/java/org/elasticsearch/index/shard/ShardPath.java

```diff
@@ -234,12 +254,10 @@ public static ShardPath selectNewPathForShard(NodeEnvironment env, ShardId shard
                         .filter((path) -> pathsToSpace.get(path).subtract(estShardSizeInBytes).compareTo(BigInteger.ZERO) > 0)
                         // Sort by the number of shards for this index
                         .sorted((p1, p2) -> {
-                                int cmp = Long.compare(pathToShardCount.getOrDefault(p1, 0L),
-                                    pathToShardCount.getOrDefault(p2, 0L));
+                                int cmp = Long.compare(p1.getNumShards(indexName), p2.getNumShards(indexName));
```

Comment:
Previously we used to sort by what shards are on disk, whereas now it's about allocated shards. This can interact with allocation of existing shards in subtle ways (as we allocate existing shards to the same data path where they were residing). Each of the two approaches has challenges, given the lack of balancing across data paths.

Reply:
Is that new shards and existing shards allocate together case I described in the above comment?